param.go

```diff
@@ -485,23 +532,85 @@ func newParamGroupedSlice(f reflect.StructField, c containerStore) (paramGrouped
 	return pg, nil
 }
 
-func (pt paramGroupedSlice) Build(c containerStore) (reflect.Value, error) {
-	var itemCount int
+func (pt paramGroupedSlice) getDecoratedValues(c containerStore) []reflect.Value {
+	var items []reflect.Value
+	stores := c.storesToRoot()
+	for _, c := range stores {
+		items = append(items, c.getDecoratedValueGroup(pt.Group, pt.Type.Elem())...)
+	}
+	return items
+}
+
+// searches the given container and its parent for matching group decorators
+// and calls them to commit values. If any of the decorators returns an error,
+// it returns that error immediately. If all decorators succeeds, it returns nil.
+func (pt paramGroupedSlice) callGroupDecorators(c containerStore) error {
+	stores := c.storesToRoot()
+	for _, c := range stores {
+		for _, d := range c.getGroupDecorators(pt.Group, pt.Type.Elem()) {
+			if err := d.Call(c); err != nil {
+				return errParamGroupFailed{
+					CtorID: d.ID(),
+					Key:    key{group: pt.Group, t: pt.Type.Elem()},
+					Reason: err,
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// searches the given container and its parent for matching group providers and
+// calls them to commit values. Returns the total number of providers it ended up
+// calling as well as an error which indicates whether any of the providers encountered
+// an error when getting called.
+func (pt paramGroupedSlice) callGroupProviders(c containerStore) (int, error) {
+	itemCount := 0
 	stores := c.storesToRoot()
 	for _, c := range stores {
 		providers := c.getGroupProviders(pt.Group, pt.Type.Elem())
 		itemCount += len(providers)
 		for _, n := range providers {
 			if err := n.Call(c); err != nil {
-				return _noValue, errParamGroupFailed{
+				return 0, errParamGroupFailed{
 					CtorID: n.ID(),
 					Key:    key{group: pt.Group, t: pt.Type.Elem()},
 					Reason: err,
 				}
 			}
 		}
 	}
+	return itemCount, nil
+}
+
+func (pt paramGroupedSlice) Build(c containerStore, decorating bool) (reflect.Value, error) {
+	// do not call this if we are already inside a decorator since
+	// it will result in an infinite recursion. (i.e. decorate -> params.BuildList() -> Decorate -> params.BuildList...)
+	// this is safe since a value can be decorated at most once in a given scope.
+	if !decorating {
+		if err := pt.callGroupDecorators(c); err != nil {
+			return _noValue, err
+		}
+	}
 
+	// Check if we have decorated values
+	decoratedItems := pt.getDecoratedValues(c)
+	if len(decoratedItems) != 0 {
+		result := reflect.MakeSlice(pt.Type, 0, len(decoratedItems))
+		for _, item := range decoratedItems {
+			result = reflect.Append(result, item)
+		}
+		return result, nil
```

Comment:
I think maybe getDecoratedValues should return an `ok bool` because an empty slice should be a valid replacement via decoration?